lib/kafka_ex/consumer_group/manager.ex

```diff
@@ -0,0 +1,476 @@
+    {:reply, state.generation_id, state}
+  end
+
+  def handle_call(:member_id, _from, state) do
+    {:reply, state.member_id, state}
+  end
+
+  def handle_call(:leader_id, _from, state) do
+    {:reply, state.leader_id, state}
+  end
+
+  def handle_call(:am_leader, _from, state) do
+    {:reply, state.leader_id && state.member_id == state.leader_id, state}
+  end
+
+  def handle_call(:assignments, _from, state) do
+    {:reply, state.assignments, state}
+  end
+
+  def handle_call(:consumer_supervisor_pid, _from, state) do
+    {:reply, state.consumer_supervisor_pid, state}
+  end
+  ######################################################################
+
+  # If `member_id` and `generation_id` aren't set, we haven't yet joined the
+  # group. `member_id` and `generation_id` are initialized by
+  # `JoinGroupResponse`.
+  def handle_info(
+    :timeout, %State{generation_id: nil, member_id: nil} = state
+  ) do
+    {:ok, new_state} = join(state)
+
+    {:noreply, new_state}
+  end
+
+  # After joining the group, a member must periodically send heartbeats to the
+  # group coordinator.
+  def handle_info(:heartbeat, %State{} = state) do
+    {:ok, new_state} = heartbeat(state)
+
+    {:noreply, new_state}
+  end
+
+  # When terminating, inform the group coordinator that this member is leaving
+  # the group so that the group can rebalance without waiting for a session
+  # timeout.
+  def terminate(_reason, %State{generation_id: nil, member_id: nil}), do: :ok
+  def terminate(_reason, %State{} = state) do
+    :ok = leave(state)
+  end
+
+  ### Helpers
+
+  # `JoinGroupRequest` is used to set the active members of a group. The
+  # response blocks until the broker has decided that it has a full list of
+  # group members. This requires that all active members send a
+  # `JoinGroupRequest`. For active members, this is triggered by the broker
+  # responding to a heartbeat with a `:rebalance_in_progress` error code. If
+  # any group members fail to send a `JoinGroupRequest` before the session
+  # timeout expires, then those group members are removed from the group and
+  # synchronization continues without them.
+  #
+  # `JoinGroupResponse` tells each member its unique member ID as well as the
+  # group's current generation ID. The broker will pick one group member to be
+  # the leader, which is reponsible for assigning partitions to all of the
+  # group members. Once a `JoinGroupResponse` is received, all group members
+  # must send a `SyncGroupRequest` (see sync/2).
+
+  defp join(
+    %State{
+      worker_name: worker_name,
+      session_timeout: session_timeout,
+      group_name: group_name,
+      topics: topics,
+      member_id: member_id
+    } = state
+  ) do
+    join_request = %JoinGroupRequest{
+      group_name: group_name,
+      member_id: member_id || "",
+      topics: topics,
+      session_timeout: session_timeout,
+    }
+
+    join_response = %JoinGroupResponse{error_code: :no_error} =
```

Comment:
This will crash if we get any error back from Kafka. I don't know if it's necessarily good to crash within a private function that's in an essentially private module. It will make it difficult to diagnose for a library user. Maybe we should crash with a helpful message here (like display a human-readable version of the error code that we got)